feat: add `cdk watch` demo content (#390)
aws-samples/aws-cdk-intro-workshop: workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md

+++
title = "CDK Watch"
weight = 300
+++

## Faster Personal Deployments

{{% notice info %}} This section is not necessary to complete the workshop, but we
recommend that you take the time to see how `cdk watch` can speed up your personal
deployments.

> {{% /notice %}}

Author: I still like where we're putting the CDK Watch stuff, but I do think it's a bit disruptive and not necessary for people to complete the workshop, so I wanted to add this disclaimer here.

Reviewer: Sure. I don't mind it 🙂.

  "watch": {
    "include": [
      "**"
    ],
    "exclude": [
      "README.md",
      "cdk*.json",
      "**/*.d.ts",
      "tsconfig.json",
      "package*.json",
      "yarn.lock",
      "node_modules",
      "test"
    ]
  },
  "context": {
    ...
  }
}
```

Now you're all set to start watching!

## See how long 'CDK Watch' takes

First, call `cdk watch`: 

```bash
cdk watch
```

This will trigger an initial deployment and immediately begin observing the files
you've specified in `cdk.json`.

For the purposes of this exercise, make sure that the initial deployment is
finished before proceeding. Otherwise, the changes you make will be queued for
deployment and released _after_ the first deployment finishes.

Let's change our lambda asset code in `lambda/hello.js` one more time. It doesn't
matter what you change it to, but let's say we want it to say
`"Good Afternoon, CDK"` now.

Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has
changed and trigger a new deployment. In this case, it will recognize that we can
hotswap the lambda asset code, so it will bypass a CloudFormation deployment and
deploy directly to the Lambda service instead.

> How fast did deployment take?

Author: Haha.. can some one actually suggest a good way to time an actual deployment. I saw @corymhall use `esbuild`, is that the general way to go?

Reviewer: I actually thought it would be awesome if we printed out, after each deployment in `cdk watch` mode, the total time the deployment took. Like we do now for the `cdk deploy` command, which ends in something like:

```
✨  Done in 78.80s.
```

Would implementing that be something you might be interested in @kaizen3031593? 🙂

Author: So `cdk deploy` doesn't actually do this I don't think. It is something that's probably useful though.

Reviewer: [It definitely does](https://www.youtube.com/watch?v=XBfgvXEaUz0&t=119s) 🙂.

Author: 😂 i was literally watching that just now and going crazy. `yarn cdk deploy` spits out the total time afterwards. but `cdk deploy` does not. I confirmed that behavior on my CDK app but can you explain why that's the case?

Author: ![Screen Shot 2021-12-29 at 3 16 04 PM](https://user-images.githubusercontent.com/36202692/147700055-0179cc59-e7e0-4f07-a328-da1dde8a26c3.png)

Reviewer: Oh, it's probably a Yarn feature then, and not a feature of the CDK CLI 😃, like I thought it was. Good catch!

Then yeah, let's add it to `cdk deploy` / `cdk watch`. I'll make an issue for it.

Reviewer: https://github.com/aws/aws-cdk/issues/18213

lambda function to respond with `"Good Morning, CDK!"` instead of `"Hello, CDK"`.

So far, it seems like the only tool we have at our disposal to update our stack is
`cdk deploy`. But `cdk deploy` takes time; it has to deploy your CloudFormation
stack and upload the `lambda` directory from your disk to the boostrap bucket. If
we're just changing our lambda code, we don't actually need to update the
CloudFormation stack, so that part of `cdk deploy` is wasted effort.

We really only need to update our lambda code. It would be great if we had
some other mechanism for doing only that...

## See how long 'CDK Deploy' takes

First, as an exercise, change your lambda code in `lambda/hello.js` slightly and 
time how long it takes to run `cdk deploy`. It will help us get a baseline for how 
long `cdk deploy` takes:

```bash
time cdk deploy
```

The output will look something like this:

```
INSERT OUTPUT HERE
```

## CDK Watch

{{% notice info %}} This command deliberately introduces drift in CloudFormation 
stacks in order to speed up deployments. For this reason, only use it for 
development purposes. Never use CDK Watch for your production deployments!

> {{% /notice %}}

Reviewer: This is actually only a feature of hotswapping, not of `cdk watch`, and you can use "watch" without hotswapping (`cdk watch --no-hotswap`).

We need to make sure the distinction between the two is clearly explained here.

Author: do we need to add more beyond what Calvin has already suggested re: `--hotswap` flag and `--no-hotswap` option?

Reviewer: I think that's enough info there.

(I'll probably change my mind once I see it all written out, but for now it's good enough 😛)

We can speed up that deployment time with `cdk watch`, which watches your
CDK application and looks for hotswappable changes. Which `cdk watch` finds a
change that can be hotswapped, it bypasses the CloudFormation deployment entirely
and instead makes the necessary SDK call to update your code for you. Changes to 
your AWS Lambda asset code, AWS Step Function State Machine definition, and other
non-resource changes have the potential to be hotswapped. Here, we will try

> `cdk watch` with your AWS Lambda asset code.

Reviewer 2: I would briefly mention the `--hotswap` flag here, and then go into more detail about how it works later on. I would change this section to something like:

"We can speed up that deploy time with `cdk watch`, which monitors your code and assets for changes and attempts to perform a deployment automatically when a change is detected. By default, `cdk watch` will use the `--hotswap` flag, which inspects the changes and determines if those changes can be hotswapped. If they can, the CDK CLI will use AWS service APIs to directly make the changes; otherwise it will fall back to performing a full CloudFormation deployment. To disable this behavior, pass `--no-hotswap` to `cdk watch`."

Reviewer: I like Calvin's suggestion!

    "include": [
      "**"
    ],
    "exclude": [
      "README.md",
      "cdk*.json",
      "**/*.d.ts",
      "**/*.js",
      "tsconfig.json",
      "package*.json",
      "yarn.lock",
      "node_modules",
      "test"
    ]
  },
  "context": {
    ...
  }
}
```

As you can see, the sample app comes with a suggested `"watch"` setting. We do in
fact want to observe our `.js` files in the `lambda` folder, so let's remove
`"**/*.js"` from the `"exclude"` list:

```json
{
  "app": "npx ts-node --prefer-ts-exts bin/workshop.ts",
  "watch": {
    "include": [
      "**"
    ],
    "exclude": [
      "README.md",
      "cdk*.json",
      "**/*.d.ts",
      "tsconfig.json",
      "package*.json",
      "yarn.lock",
      "node_modules",
      "test"
    ]
  },
  "context": {
    ...
  }
}
```

Now you're all set to start watching!

## See how long 'CDK Watch' takes

First, call `cdk watch`: 

```bash
cdk watch
```

This will trigger an initial deployment and immediately begin observing the files
you've specified in `cdk.json`.

For the purposes of this exercise, make sure that the initial deployment is
finished before proceeding. Otherwise, the changes you make will be queued for

> deployment and released _after_ the first deployment finishes.

Reviewer: Not sure we need this paragraph - even if the changes come during, they will still be picked up, so I don't think it makes sense to add this disclaimer.

Let's change our lambda asset code in `lambda/hello.js` one more time. It doesn't
matter what you change it to, but let's say we want it to say
`"Good Afternoon, CDK"` now.

Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has
changed and trigger a new deployment. In this case, it will recognize that we can
hotswap the lambda asset code, so it will bypass a CloudFormation deployment and
deploy directly to the Lambda service instead.

How fast did deployment take?

```
INSERT PHOTO HERE

> ```

Reviewer 2: I would modify this a bit, something like:

"Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has changed and trigger a new deployment. 

```
INSERT PHOTO HERE
```

The warning message indicates that this lambda code change is hotswappable, so it will bypass CloudFormation and go directly to the Lambda Service API. This introduces drift between your CloudFormation template and your deployed application code. Because of this drift, hotswapping should never be used in a production environment. Hotswap deployments are faster, but lack the robust safety features of CloudFormation deployments. To disable hotswapping, pass the `--no-hotswap` flag to `cdk watch`."

Reviewer: I feel like we addressed this already above, with hotswapping. I don't think it makes sense to repeat all of that here.

It's a good idea to mention `cdk watch --no-hotswap` though.

+++
title = "CDK Watch"
weight = 300
+++

> ## Faster Personal Deployment

Reviewer: I don't think we need to uppercase all letters in the title here:

```suggestion
## Faster personal deployments
```

{{% notice info %}} This section is not necessary to complete the workshop, but we
recommend that you take the time to see how `cdk deploy --hotswap` and `cdk watch` 
can speed up your personal deployments.
{{% /notice %}}

It's great that we have a working lambda! But what if you want to tweak the lambda
code to get it just right? Let's say that we have now decided that we want our
lambda function to respond with `"Good Morning, CDK!"` instead of `"Hello, CDK"`.

So far, it seems like the only tool we have at our disposal to update our stack is
`cdk deploy`. But `cdk deploy` takes time; it has to deploy your CloudFormation
stack and upload the `lambda` directory from your disk to the boostrap bucket. If
we're just changing our lambda code, we don't actually need to update the
CloudFormation stack, so that part of `cdk deploy` is wasted effort.

We really only need to update our lambda code. It would be great if we had
some other mechanism for doing only that...

> ## See how long 'CDK Deploy' takes

Reviewer: ```suggestion
## Timing `cdk deploy`
```

First, as an exercise, change your lambda code in `lambda/hello.js` slightly and 
time how long it takes to run `cdk deploy`. It will help us get a baseline for how 
long `cdk deploy` takes:

```bash
time cdk deploy
```

The output will look something like this:

```
INSERT OUTPUT HERE
```

> ## CDK Deploy Hotswap

Reviewer 2: I'm not sure I like the title, maybe something like `## Performing a Hotswap Deployment`? `## CDK Deploy --hotswap`? `Hotswap Deployments`?

Reviewer: I think `## Hotswap deployments` is a good title.

{{% notice info %}} This command deliberately introduces drift in CloudFormation 
stacks in order to speed up deployments. For this reason, only use it for 
development purposes. Never use hotswap for your production deployments!
{{% /notice %}}

We can speed up that deployment time with `cdk deploy --hotswap`, which will
assess whether a hotswap deployment can be performed instead of a CloudFormation
deployment. If possible, the CDK CLI will use AWS service APIs to directly make
the changes; otherwise it will fall back to performing a full CloudFormation
deployment.

> Here, we will use `cdk deploy --hotswap` to detect a hotswappable change to your 

Reviewer 2: ```suggestion
Here, we will use `cdk deploy --hotswap` to deploy a hotswappable change to your 
```

Reviewer: +1

AWS Lambda asset code.

> ## See how long 'CDK Deploy --Hotswap' takes

Reviewer: ```suggestion
## Timing `cdk deploy --hotswap`
```

`"exclude"`, each of which can be either a single string or an array of strings.
Each entry is interpreted as a path relative to the location of the `cdk.json`
file. Globs, both `*` and `**`, are allowed to be used.

Your `cdk.json` file should look like this:

```json
{
  "app": "npx ts-node --prefer-ts-exts bin/cdk-workshop.ts",
  "watch": {
    "include": [
      "**"
    ],
    "exclude": [
      "README.md",
      "cdk*.json",
      "**/*.d.ts",
      "**/*.js",
      "tsconfig.json",
      "package*.json",
      "yarn.lock",
      "node_modules",
      "test"
    ]
  },
  "context": {
    // ...
  }
}
```

As you can see, the sample app comes with a suggested `"watch"` setting. We do in
fact want to observe our `.js` files in the `lambda` folder, so let's remove
`"**/*.js"` from the `"exclude"` list:

```json
{
  "app": "npx ts-node --prefer-ts-exts bin/cdk-workshop.ts",
  "watch": {
    "include": [
      "**"
    ],
    "exclude": [
      "README.md",
      "cdk*.json",
      "**/*.d.ts",
      "tsconfig.json",
      "package*.json",
      "yarn.lock",
      "node_modules",
      "test"
    ]
  },
  "context": {
    // ...
  }
}
```

Now you're all set to start watching!

> ## See how long 'CDK Watch' takes

Reviewer: ```suggestion
## Timing `cdk watch`
```

First, call `cdk watch`: 

```bash
cdk watch
```

This will trigger an initial deployment and immediately begin observing the files
you've specified in `cdk.json`.

Let's change our lambda asset code in `lambda/hello.js` one more time. It doesn't
matter what you change it to, but let's say we want it to say

> `"Good Afternoon, CDK"` now.

Reviewer: Let's be more specific here maybe? I feel like just saying "make any change" each time doesn't "jive" well with the rest of the workshop.

How about we say to make a specific change each time, and we use the nice highlighting feature to show clearly in the example what line(s) did we change each time?

Author: the highlighting feature is cool, didn't even realize how we did that.

> Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has

Reviewer: Let's just say "Once you save the changes to your Lambda code file", or something.

+++
title = "CDK Watch"
weight = 300
+++

## Faster personal deployments

{{% notice info %}} This section is not necessary to complete the workshop, but we
recommend that you take the time to see how `cdk deploy --hotswap` and `cdk watch` 
can speed up your personal deployments.
{{% /notice %}}

> It's great that we have a working lambda! But what if you want to tweak the lambda

Reviewer: Let's us "we" consistently:

```suggestion
It's great that we have a working lambda! But what if we want to tweak the lambda
```

## Timing `cdk deploy --hotswap`

Let's change the lambda code in `lambda/hello.js` another time:

{{<highlight js "hl_lines=6">}}
exports.handler = async function(event) {
  console.log("request:", JSON.stringify(event, undefined, 2));
  return {
    statusCode: 200,
    headers: { "Content-Type": "text/plain" },
    body: `Good Afternoon, CDK! You've hit ${event.path}\n`
  };
};
{{</highlight>}}

Now, let's run `cdk deploy --hotswap`:

```
cdk deploy --hotswap
```

The output will look something like this:

```
✨  Synthesis time: 6.44s

⚠️ The --hotswap flag deliberately introduces CloudFormation drift to speed up deployments
⚠️ It should only be used for development - never use it for your production Stacks!

CdkWorkshopStack: deploying...
✨ hotswapping resources:
   ✨ Lambda Function 'CdkWorkshopStack-HelloHandler2E4FBA4D-tEZTcXqG8YYe'
✨ Lambda Function 'CdkWorkshopStack-HelloHandler2E4FBA4D-tEZTcXqG8YYe' hotswapped!

 ✅  CdkWorkshopStack

✨  Deployment time: 3.07s

Stack ARN:
arn:aws:cloudformation:REGION:ACCOUNT-ID:stack/CdkWorkshopStack/STACK-ID

✨  Total time: 9.51s
```

Wow, deploying a hotswapped change took 3 seconds, while a full deployment took 67 seconds! 

> But take a look and read the warning message thoroughly - it's important!:

Reviewer: I think we don't need the colon here, the exclamation mark is good enough:

```suggestion
But take a look and read the warning message thoroughly - it's important!
```

```
⚠️ The --hotswap flag deliberately introduces CloudFormation drift to speed up deployments
⚠️ It should only be used for development - never use it for your production Stacks!
```

## Did the code actually change?

Wow that was fast. Did the code actually change? Let's go to the AWS Lambda Console and
double check!

1. Open the [AWS Lambda
   Console](https://console.aws.amazon.com/lambda/home#/functions) (make sure
   you are in the correct region).

    You should see our function:

    ![](./lambda-1.png)

2. Click on the function name to go to the console.

3. The code should be loaded onto the screen. Did your change show up?

>     ![](./lambda-5.png)

Reviewer: You might consider also testing that the function changed by calling it, and observing the result it returned. That's super minor though, this is fine too if you want to keep it as-is.

Author: We do exactly this (test the result) in the previous page where we set up the lambda. So I think it is fine to keep it as-is here; feels like observing the code change is evidence enough.

## CDK Watch

We can do better than calling `cdk deploy` or `cdk deploy --hotswap` each time.
`cdk watch` is similar to `cdk deploy` except that instead of being a one-shot
operation, it monitors your code and assets for changes and attempts to perform a 
deployment automatically when a change is detected. By default, `cdk watch` will 
use the `--hotswap` flag, which inspects the changes and determines if those 
changes can be hotswapped. To disable this behavior, you can call
`cdk watch --no-hotswap`.

Once we set it up, we can use `cdk watch` to detect both hotswappable changes and
changes that require full CloudFormation deployment.

## Modify your `cdk.json` file

> When the `cdk watch` command runs, the files that it observes is determined by the

Reviewer: ```suggestion
When the `cdk watch` command runs, the files that it observes are determined by the
```

Author: good catch :)

`"watch"` setting in the `cdk.json` file. It has two sub-keys, `"include"` and
`"exclude"`, each of which can be either a single string or an array of strings.
Each entry is interpreted as a path relative to the location of the `cdk.json`
file. Globs, both `*` and `**`, are allowed to be used.

> Your `cdk.json` file should look like this:

Reviewer: ```suggestion
Your `cdk.json` file should look similar to this:
```
